transformer_engine/pytorch/optimizers/fused_adam.py

```diff
@@ -172,45 +211,170 @@ def step(self, closure=None, grad_scaler=None):
             p_16_master = []
             p_32_master = []
 
-            for p, p_master in zip(group["params"], group_master["params"]):
-                if p.grad is None:
-                    continue
-                if p.grad.data.is_sparse:
-                    raise RuntimeError("FusedAdam does not support sparse gradients.")
-
-                state = self.state[p]
-                # State initialization
-                if len(state) == 0:
-                    # Exponential moving average of gradient values
-                    state["exp_avg"] = torch.zeros_like(p.data).float()
-                    # Exponential moving average of squared gradient values
-                    state["exp_avg_sq"] = torch.zeros_like(p.data).float()
-
-                if p.dtype == torch.float16:
-                    if self.master_weights:
-                        p_16_master.append(p_master.data)
-                    g_16.append(p.grad.data)
-                    p_16.append(p.data)
-                    m_16.append(state["exp_avg"])
-                    v_16.append(state["exp_avg_sq"])
-                elif p.dtype == torch.bfloat16:
-                    g_bf.append(p.grad)
-                    p_bf.append(p)
-                    m_bf.append(state["exp_avg"])
-                    v_bf.append(state["exp_avg_sq"])
-                elif p.dtype == torch.float32:
-                    if self.master_weights:
-                        p_32_master.append(p_master.data)
-                    g_32.append(p.grad.data)
-                    p_32.append(p.data)
-                    m_32.append(state["exp_avg"])
-                    v_32.append(state["exp_avg_sq"])
-                else:
-                    raise RuntimeError("FusedAdam only support fp16 and fp32.")
+            # when fuse_dtype_casting is True, the adam kernel also needs to update extra parameters.
+            p_main_of_fp8 = []
+            p_main_of_f16 = []
+            g_main_of_fp8 = []
+            g_main_of_f16 = []
+            m_main_of_fp8 = []
+            m_main_of_f16 = []
+            v_main_of_fp8 = []
+            v_main_of_f16 = []
+            p_extra_fp16_out = []
+            p_extra_fp8_out = []
+            scales = []
+            amaxes = []
```

Comment:
It seems like we are duplicating some logic from the non-FP8 case. Also, we are not handling the case where we have a mix of FP8, FP16, BF16, and FP32 params.

Reply:
Yes. There is some redundant logic here, which is also due to my reluctance to disrupt existing functionality.
The cases that can currently be handled should include:
1. master_weight: FP32/FP16/BF16, model weight: FP16/BF16
2. master_weight: FP32/FP16/BF16, model weight: FP16/BF16 + FP8

I guess that FP32+FP8 model weights are unlikely to be used in practice, so I haven't considered this scenario.

Reply:
I have refactored this part of the code and simplified the logic.

Reply:
I see that we simply error out if there are FP32 params. We will also produce incorrect values is there is a mix of FP16 and BF16 params. It's a weird case, so I'm fine if we just add a check in the constructor and throw an error if there's a mix of FP16 and BF16.